{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "82445ade-bd78-434f-a921-91a3e6dd51fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import subprocess\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26097f1d-4ba4-45f0-83eb-562685966bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path with files outputted from bedtools intersect\n",
    "PATH = \"/cluster/scratch/taekim/data_oxidation/gene_bodies_intersect\" \n",
    "OUTPUT_FILE = \"../data_normalized/gene_bodies_Normalized.csv\"\n",
    "GENOME_PATH = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Navnit/genomes/mouse/GRCm39_NCBI_Bowtie2.fasta\"\n",
    "# Original path with all CpG coordinates\n",
    "ORG_PATH = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/knownGenes_canononical_GRCm39_GENCODE.VM36.bed\" \n",
    "\n",
    "# Define column names for the input files\n",
    "INPUT_COLUMNS = [\"Chr1\", \"Start1\", \"End1\", \"Value\", \"MAPQ\", \"Chr2\", \"Start2\", \"End2\", \"Gene\", \"idk\", \"ignore\"]\n",
    "\n",
    "# Define columns for the output file\n",
    "OUTPUT_COLUMNS = [\"id\", \"sample\", \"gene\", \"strand\", \"GC_count\", \"damage\", \n",
    "                  \"GC_normalized_damage\", \"median\", \"median_normalized_damage\"]\n",
    "\n",
    "# Define the median values for each sample\n",
    "# Median value for each sample from 100kb bins\n",
    "median_values = {\n",
    "    \"Sample_14_CRS_evening_S14_\": 2.546314,\n",
    "    \"Sample_15_Ctrl_evening_S15_\": 3.570246,\n",
    "    \"Sample_05_Ctrl_morning_S5_\": 6.184096,\n",
    "    \"Sample_01_Ctrl_morning_S1_\": 6.921409,\n",
    "    \"Sample_16_CRS_evening_S16_\": 2.879618,\n",
    "    \"Sample_11_Ctrl_evening_S11_\": 2.485161,\n",
    "    \"Sample_13_Ctrl_evening_S13_\": 2.403964,\n",
    "    \"Sample_08_CRS_morning_S8_\": 5.246539,\n",
    "    \"Sample_20_CRS_evening_S20_\": 10.633666,\n",
    "    \"Sample_18_CRS_evening_S18_\": 4.602043,\n",
    "    \"Sample_17_Ctrl_evening_S17_\": 2.929046,\n",
    "    \"Sample_04_CRS_morning_S4_\": 6.227455,\n",
    "    \"Sample_19_Ctrl_evening_S19_\": 2.545900,\n",
    "    \"Sample_03_Ctrl_morning_S3_\": 9.529751,\n",
    "    \"Sample_02_CRS_morning_S2_\": 7.151634,\n",
    "    \"Sample_09_Ctrl_morning_S9_\": 7.923174,\n",
    "    \"Sample_10_CRS_morning_S10_\": 8.815860,\n",
    "    \"Sample_06_CRS_morning_S6_\": 7.930738,\n",
    "    \"Sample_12_CRS_evening_S12_\": 2.849694,\n",
    "    \"Sample_07_Ctrl_morning_S7_\": 4.924262\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4217a79c-cc57-4af1-ba70-9c55bf81cadc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: /cluster/software/stacks/2024-06/spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-12.2.0/bedtools2-2.31.0-a4obbslkxntgdx2criopqpwx662gcftq/bin/bedtools getfasta -fi /nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Navnit/genomes/mouse/GRCm39_NCBI_Bowtie2.fasta -bed /nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/knownGenes_canononical_GRCm39_GENCODE.VM36.bed -bedOut > /cluster/scratch/taekim/data_oxidation/gene_bodies_intersect/gene_bodies.SEQ.bed\n",
      "Command completed successfully\n"
     ]
    }
   ],
   "source": [
    "def run_bedtools_getfasta(bed_file, output_file, genome_path):\n",
    "    \"\"\"Run bedtools getfasta command to get sequence data for a BED file\"\"\"\n",
    "    command = f\"/cluster/software/stacks/2024-06/spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-12.2.0/bedtools2-2.31.0-a4obbslkxntgdx2criopqpwx662gcftq/bin/bedtools getfasta -fi {genome_path} -bed {bed_file} -bedOut > {output_file}\"\n",
    "    print(f\"Running command: {command}\")\n",
    "    \n",
    "    try:\n",
    "        result = subprocess.run(command, shell=True, check=True, \n",
    "                               stdout=subprocess.PIPE, stderr=subprocess.PIPE,\n",
    "                               text=True)\n",
    "        print(\"Command completed successfully\")\n",
    "        return True\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running bedtools: {e}\")\n",
    "        print(f\"STDERR: {e.stderr}\")\n",
    "        return False\n",
    "\n",
    "SEQ_COLUMNS = [\"Chr\", \"Start\", \"End\", \"Gene\", \"idk\", \"ignore\", \"Seq\"]\n",
    "seq_output_file = os.path.join(PATH, \"gene_bodies.SEQ.bed\")\n",
    "# Run getfasta with original CpG coordinate file\n",
    "run_bedtools_getfasta(ORG_PATH, seq_output_file, GENOME_PATH)\n",
    "seq_df = pd.read_csv(seq_output_file, sep=\"\\t\", header=None, names=SEQ_COLUMNS)\n",
    "\n",
    "seq_counts_plus = seq_df.copy()\n",
    "seq_counts_plus['id'] = seq_counts_plus['Gene'] + '+'  # Use Gene for consistency with previous code\n",
    "seq_counts_plus['GC_count'] = seq_counts_plus['Seq'].str.upper().str.count('G')\n",
    "seq_counts_plus = seq_counts_plus.drop_duplicates('id')\n",
    "\n",
    "seq_counts_minus = seq_df.copy()\n",
    "seq_counts_minus['id'] = seq_counts_minus['Gene'] + '-'  # Use Gene for consistency with previous code\n",
    "seq_counts_minus['GC_count'] = seq_counts_minus['Seq'].str.upper().str.count('C')\n",
    "seq_counts_minus = seq_counts_minus.drop_duplicates('id')\n",
    "\n",
    "# Combine both strand data\n",
    "seq_counts = pd.concat([seq_counts_plus, seq_counts_minus])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c5c6ed40-ed50-40cf-9d7f-435a0eadabc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First pass: Collecting all unique IDs...\n",
      "Collected 152236 unique IDs across all samples\n"
     ]
    }
   ],
   "source": [
    "# Collect all files for each sample\n",
    "bed_files = glob.glob(os.path.join(PATH, \"*strand.bed\"))\n",
    "\n",
    "all_unique_ids = set()  # To store all unique IDs across all samples\n",
    "# First pass: Collect all unique IDs across all samples\n",
    "print(\"First pass: Collecting all unique IDs...\")\n",
    "for file_path in sorted(bed_files):\n",
    "    file_name = os.path.basename(file_path)\n",
    "    try:\n",
    "        # Read the file\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS,  on_bad_lines='skip')\n",
    "\n",
    "\n",
    "        if \"plus_strand\" in file_name:\n",
    "            strand = \"+\"\n",
    "        elif \"minus_strand\" in file_name:\n",
    "            strand = \"-\"        \n",
    "        # Create the identifier\n",
    "        ids = df['Gene'] + strand\n",
    "        \n",
    "        # Add to the set of all unique IDs\n",
    "        all_unique_ids.update(ids.unique())\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"  Error collecting IDs from {file_name}: {str(e)}\")\n",
    "\n",
    "print(f\"Collected {len(all_unique_ids)} unique IDs across all samples\")\n",
    "\n",
    "\n",
    "# Create a dictionary to store chromosome info for each ID\n",
    "id_to_chrom = {}\n",
    "for file_path in sorted(bed_files):\n",
    "    file_name = os.path.basename(file_path)\n",
    "    try:\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS)\n",
    "        # Map each ID to its chromosome\n",
    "        id_df = df.copy()\n",
    "        \n",
    "        if \"plus_strand\" in file_name:\n",
    "            strand = \"+\"\n",
    "        elif \"minus_strand\" in file_name:\n",
    "            strand = \"-\"   \n",
    "            \n",
    "        id_df['id'] = id_df['Gene'] + strand\n",
    "        \n",
    "        for _, row in id_df.drop_duplicates('id').iterrows():\n",
    "            id_to_chrom[row['id']] = row['Chr1']\n",
    "    except Exception as e:\n",
    "        print(f\"  Error mapping chromosomes: {str(e)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3862cee5-bcc8-4547-a834-be4e82e0f1df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting processing of samples\n",
      "starting\n",
      "Processing file: Sample_01_Ctrl_morning_S1_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3834667\n",
      "File data: 59881\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      47  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      -      8786              0.005349   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                   0.77288  \n",
      "Processing file: Sample_01_Ctrl_morning_S1_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 3075012\n",
      "File data: 59867\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      56  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      +      7402              0.007566   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  1.093061  \n",
      "Processing file: Sample_02_CRS_morning_S2_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3492953\n",
      "File data: 60152\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      65  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      -      8786              0.007398  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.034468  \n",
      "Processing file: Sample_02_CRS_morning_S2_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 3248807\n",
      "File data: 60298\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      37  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      +      7402              0.004999  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.698952  \n",
      "Processing file: Sample_03_Ctrl_morning_S3_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 4724323\n",
      "File data: 63205\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      84  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      -      8786              0.009561   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  1.003244  \n",
      "Processing file: Sample_03_Ctrl_morning_S3_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 4394782\n",
      "File data: 63122\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      +      7402              0.007025   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  0.737179  \n",
      "Processing file: Sample_04_CRS_morning_S4_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3375749\n",
      "File data: 59296\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      51  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      -      8786              0.005805  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.932113  \n",
      "Processing file: Sample_04_CRS_morning_S4_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 2950542\n",
      "File data: 59406\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      28  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      +      7402              0.003783  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.607433  \n",
      "Processing file: Sample_05_Ctrl_morning_S5_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3829381\n",
      "File data: 58409\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      -      8786              0.005919   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  0.957053  \n",
      "Processing file: Sample_05_Ctrl_morning_S5_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 3288285\n",
      "File data: 58342\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      36  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      +      7402              0.004864   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  0.786461  \n",
      "Processing file: Sample_06_CRS_morning_S6_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 4132065\n",
      "File data: 62167\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      60  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      -      8786              0.006829  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.861086  \n",
      "Processing file: Sample_06_CRS_morning_S6_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 3795624\n",
      "File data: 62124\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      48  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      +      7402              0.006485  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.817671  \n",
      "Processing file: Sample_07_Ctrl_morning_S7_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3136719\n",
      "File data: 56519\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      63  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      -      8786               0.00717   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  1.456157  \n",
      "Processing file: Sample_07_Ctrl_morning_S7_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 2636361\n",
      "File data: 56394\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      31  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      +      7402              0.004188   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  0.850494  \n",
      "Processing file: Sample_08_CRS_morning_S8_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 2835468\n",
      "File data: 58042\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      -      8786              0.005919  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.128078  \n",
      "Processing file: Sample_08_CRS_morning_S8_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 2469385\n",
      "File data: 58021\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      40  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      +      7402              0.005404  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.030002  \n",
      "Processing file: Sample_09_Ctrl_morning_S9_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3125205\n",
      "File data: 60791\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      71  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      -      8786              0.008081   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.019924  \n",
      "Processing file: Sample_09_Ctrl_morning_S9_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 2891032\n",
      "File data: 60530\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      51  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      +      7402               0.00689   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  0.869605  \n",
      "Processing file: Sample_10_CRS_morning_S10_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 3283968\n",
      "File data: 63287\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      66  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      -      8786              0.007512  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.852095  \n",
      "Processing file: Sample_10_CRS_morning_S10_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 3152518\n",
      "File data: 63418\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      49  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      +      7402               0.00662  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                    0.7509  \n",
      "Processing file: Sample_11_Ctrl_evening_S11_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1402657\n",
      "File data: 49804\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      24  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      -      8786              0.002732   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.099172  \n",
      "Processing file: Sample_11_Ctrl_evening_S11_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1125614\n",
      "File data: 49852\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      13  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      +      7402              0.001756   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  0.706708  \n",
      "Processing file: Sample_12_CRS_evening_S12_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1584874\n",
      "File data: 51939\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      34  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      -      8786               0.00387   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  1.357968  \n",
      "Processing file: Sample_12_CRS_evening_S12_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1273887\n",
      "File data: 52055\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      16  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      +      7402              0.002162   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                   0.75853  \n",
      "Processing file: Sample_13_Ctrl_evening_S13_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1810619\n",
      "File data: 48958\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      17  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      -      8786              0.001935   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  0.804877  \n",
      "Processing file: Sample_13_Ctrl_evening_S13_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1266753\n",
      "File data: 48906\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      26  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      +      7402              0.003513   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  1.461155  \n",
      "Processing file: Sample_14_CRS_evening_S14_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1106778\n",
      "File data: 49672\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      25  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      -      8786              0.002845   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  1.117473  \n",
      "Processing file: Sample_14_CRS_evening_S14_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1011383\n",
      "File data: 49488\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      18  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      +      7402              0.002432   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  0.955018  \n",
      "Processing file: Sample_15_Ctrl_evening_S15_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 2215131\n",
      "File data: 55756\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      37  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      -      8786              0.004211   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  1.179539  \n",
      "Processing file: Sample_15_Ctrl_evening_S15_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1723110\n",
      "File data: 55879\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      23  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      +      7402              0.003107   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  0.870323  \n",
      "Processing file: Sample_16_CRS_evening_S16_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1415521\n",
      "File data: 52018\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      24  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      -      8786              0.002732   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.948604  \n",
      "Processing file: Sample_16_CRS_evening_S16_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1231670\n",
      "File data: 52247\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      21  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      +      7402              0.002837   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.985225  \n",
      "Processing file: Sample_17_Ctrl_evening_S17_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1650034\n",
      "File data: 51560\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      13  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      -      8786               0.00148   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  0.505157  \n",
      "Processing file: Sample_17_Ctrl_evening_S17_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1317755\n",
      "File data: 51670\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      22  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      +      7402              0.002972   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  1.014723  \n",
      "Processing file: Sample_18_CRS_evening_S18_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1899518\n",
      "File data: 57265\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      49  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      -      8786              0.005577   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  1.211865  \n",
      "Processing file: Sample_18_CRS_evening_S18_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1802835\n",
      "File data: 57365\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      27  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      +      7402              0.003648   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  0.792618  \n",
      "Processing file: Sample_19_Ctrl_evening_S19_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1401357\n",
      "File data: 50389\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      29  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      -      8786              0.003301  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.296479  \n",
      "Processing file: Sample_19_Ctrl_evening_S19_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1112379\n",
      "File data: 50230\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      19  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      +      7402              0.002567  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.008238  \n",
      "Processing file: Sample_20_CRS_evening_S20_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 4779373\n",
      "File data: 67231\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-     100  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      -      8786              0.011382   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                   1.07035  \n",
      "Processing file: Sample_20_CRS_evening_S20_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 4182197\n",
      "File data: 67351\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      92  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      +      7402              0.012429   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.168842  \n"
     ]
    }
   ],
   "source": [
    "# Create an empty DataFrame for the final output\n",
    "all_data = pd.DataFrame(columns=OUTPUT_COLUMNS)\n",
    "\n",
    "print(\"Starting processing of samples\")\n",
    "\n",
    "# Process each file\n",
    "sample_data_dict = {}  # Store data for each sample\n",
    "\n",
    "print(\"starting\")\n",
    "for file_path in sorted(bed_files):\n",
    "    file_name = os.path.basename(file_path)\n",
    "    print(f\"Processing file: {file_name}\")\n",
    "    \n",
    "    # Extract sample name and determine strand\n",
    "    sample_name = file_name.split('.')[0]\n",
    "    \n",
    "    if \"plus_strand\" in file_name:\n",
    "        strand = \"+\"\n",
    "        nucleotide = 'G'  # Count G on plus strand\n",
    "    elif \"minus_strand\" in file_name:\n",
    "        strand = \"-\"\n",
    "        nucleotide = 'C'  # Count C on minus strand\n",
    "    else:\n",
    "        print(f\"  Warning: Could not determine strand for {file_name}\")\n",
    "        continue\n",
    "    \n",
    "    try:\n",
    "        # Read and process the file\n",
    "        # dataframe of intersection from oxidation sites per sample and cpg coordinates\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS)\n",
    "        print(f\"  Records: {len(df)}\")\n",
    "        \n",
    "        # Create IDs and calculate damage per genomic position\n",
    "        df['id'] = df['Gene'] + strand \n",
    "        \n",
    "        # Create damage_df with both summed damage and first gene in one operation\n",
    "        damage_df = df.groupby('id').agg({\n",
    "            'Value': 'sum',\n",
    "            'Gene': 'first',\n",
    "            'Chr1': 'first'\n",
    "        }).reset_index()\n",
    "        \n",
    "        # Rename Value column to damage\n",
    "        damage_df = damage_df.rename(columns={'Value': 'damage'})\n",
    "        damage_df = damage_df.rename(columns={'Gene': 'gene'})\n",
    "        damage_df = damage_df.rename(columns={'Chr1': 'chromosome'})\n",
    "        \n",
    "        # Create base result dataframe\n",
    "        file_data = damage_df.copy()\n",
    "        file_data['sample'] = sample_name\n",
    "        file_data['strand'] = strand\n",
    "        \n",
    "        # 4. Merge the counts with our result dataframe\n",
    "        file_data = file_data.merge(\n",
    "            seq_counts[['id', 'GC_count']], \n",
    "            on='id', \n",
    "            how='left'\n",
    "        )\n",
    "        \n",
    "        # Calculate normalized damage\n",
    "        file_data['GC_normalized_damage'] = file_data['damage'] / file_data['GC_count'].replace(0, np.nan)\n",
    "        \n",
    "        # Add median normalization if available\n",
    "        if sample_name in median_values:\n",
    "            file_data['median'] = median_values[sample_name]\n",
    "            file_data['median_normalized_damage'] = 1000 * file_data['GC_normalized_damage'] / file_data['median']\n",
    "        else:\n",
    "            print(f\"  Warning: No median value found for sample {sample_name}\")\n",
    "            file_data['median'] = np.nan\n",
    "            file_data['median_normalized_damage'] = np.nan\n",
    "        \n",
    "        # Store processed data\n",
    "        # file_data saved per sample\n",
    "        sample_data_dict[f\"{sample_name}_{strand}\"] = file_data\n",
    "        print(f\"File data: {len(file_data)}\")\n",
    "        print (file_data.head(1))\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"  Error processing {file_name}: {str(e)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6b34f245-69af-449e-be0c-b9d8f698cfd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding missing IDs to each sample...\n",
      "Sample_01_Ctrl_morning_S1__-\n",
      "There are 59881 unique ids\n",
      "  Adding 92355 missing IDs to sample Sample_01_Ctrl_morning_S1_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      47  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      -      8786              0.005349   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                   0.77288  \n",
      "Sample_01_Ctrl_morning_S1__+\n",
      "There are 59867 unique ids\n",
      "  Adding 92369 missing IDs to sample Sample_01_Ctrl_morning_S1_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      56  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      +      7402              0.007566   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  1.093061  \n",
      "Sample_02_CRS_morning_S2__-\n",
      "There are 60152 unique ids\n",
      "  Adding 92084 missing IDs to sample Sample_02_CRS_morning_S2_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      65  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      -      8786              0.007398  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.034468  \n",
      "Sample_02_CRS_morning_S2__+\n",
      "There are 60298 unique ids\n",
      "  Adding 91938 missing IDs to sample Sample_02_CRS_morning_S2_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      37  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      +      7402              0.004999  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.698952  \n",
      "Sample_03_Ctrl_morning_S3__-\n",
      "There are 63205 unique ids\n",
      "  Adding 89031 missing IDs to sample Sample_03_Ctrl_morning_S3_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      84  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      -      8786              0.009561   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  1.003244  \n",
      "Sample_03_Ctrl_morning_S3__+\n",
      "There are 63122 unique ids\n",
      "  Adding 89114 missing IDs to sample Sample_03_Ctrl_morning_S3_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      +      7402              0.007025   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  0.737179  \n",
      "Sample_04_CRS_morning_S4__-\n",
      "There are 59296 unique ids\n",
      "  Adding 92940 missing IDs to sample Sample_04_CRS_morning_S4_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      51  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      -      8786              0.005805  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.932113  \n",
      "Sample_04_CRS_morning_S4__+\n",
      "There are 59406 unique ids\n",
      "  Adding 92830 missing IDs to sample Sample_04_CRS_morning_S4_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      28  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      +      7402              0.003783  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.607433  \n",
      "Sample_05_Ctrl_morning_S5__-\n",
      "There are 58409 unique ids\n",
      "  Adding 93827 missing IDs to sample Sample_05_Ctrl_morning_S5_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      -      8786              0.005919   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  0.957053  \n",
      "Sample_05_Ctrl_morning_S5__+\n",
      "There are 58342 unique ids\n",
      "  Adding 93894 missing IDs to sample Sample_05_Ctrl_morning_S5_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      36  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      +      7402              0.004864   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  0.786461  \n",
      "Sample_06_CRS_morning_S6__-\n",
      "There are 62167 unique ids\n",
      "  Adding 90069 missing IDs to sample Sample_06_CRS_morning_S6_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      60  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      -      8786              0.006829  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.861086  \n",
      "Sample_06_CRS_morning_S6__+\n",
      "There are 62124 unique ids\n",
      "  Adding 90112 missing IDs to sample Sample_06_CRS_morning_S6_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      48  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      +      7402              0.006485  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.817671  \n",
      "Sample_07_Ctrl_morning_S7__-\n",
      "There are 56519 unique ids\n",
      "  Adding 95717 missing IDs to sample Sample_07_Ctrl_morning_S7_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      63  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      -      8786               0.00717   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  1.456157  \n",
      "Sample_07_Ctrl_morning_S7__+\n",
      "There are 56394 unique ids\n",
      "  Adding 95842 missing IDs to sample Sample_07_Ctrl_morning_S7_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      31  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      +      7402              0.004188   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  0.850494  \n",
      "Sample_08_CRS_morning_S8__-\n",
      "There are 58042 unique ids\n",
      "  Adding 94194 missing IDs to sample Sample_08_CRS_morning_S8_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      52  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      -      8786              0.005919  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.128078  \n",
      "Sample_08_CRS_morning_S8__+\n",
      "There are 58021 unique ids\n",
      "  Adding 94215 missing IDs to sample Sample_08_CRS_morning_S8_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      40  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      +      7402              0.005404  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.030002  \n",
      "Sample_09_Ctrl_morning_S9__-\n",
      "There are 60791 unique ids\n",
      "  Adding 91445 missing IDs to sample Sample_09_Ctrl_morning_S9_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      71  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      -      8786              0.008081   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.019924  \n",
      "Sample_09_Ctrl_morning_S9__+\n",
      "There are 60530 unique ids\n",
      "  Adding 91706 missing IDs to sample Sample_09_Ctrl_morning_S9_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      51  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      +      7402               0.00689   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  0.869605  \n",
      "Sample_10_CRS_morning_S10__-\n",
      "There are 63287 unique ids\n",
      "  Adding 88949 missing IDs to sample Sample_10_CRS_morning_S10_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      66  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      -      8786              0.007512  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.852095  \n",
      "Sample_10_CRS_morning_S10__+\n",
      "There are 63418 unique ids\n",
      "  Adding 88818 missing IDs to sample Sample_10_CRS_morning_S10_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      49  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      +      7402               0.00662  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                    0.7509  \n",
      "Sample_11_Ctrl_evening_S11__-\n",
      "There are 49804 unique ids\n",
      "  Adding 102432 missing IDs to sample Sample_11_Ctrl_evening_S11_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      24  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      -      8786              0.002732   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.099172  \n",
      "Sample_11_Ctrl_evening_S11__+\n",
      "There are 49852 unique ids\n",
      "  Adding 102384 missing IDs to sample Sample_11_Ctrl_evening_S11_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      13  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      +      7402              0.001756   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  0.706708  \n",
      "Sample_12_CRS_evening_S12__-\n",
      "There are 51939 unique ids\n",
      "  Adding 100297 missing IDs to sample Sample_12_CRS_evening_S12_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      34  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      -      8786               0.00387   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  1.357968  \n",
      "Sample_12_CRS_evening_S12__+\n",
      "There are 52055 unique ids\n",
      "  Adding 100181 missing IDs to sample Sample_12_CRS_evening_S12_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      16  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      +      7402              0.002162   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                   0.75853  \n",
      "Sample_13_Ctrl_evening_S13__-\n",
      "There are 48958 unique ids\n",
      "  Adding 103278 missing IDs to sample Sample_13_Ctrl_evening_S13_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      17  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      -      8786              0.001935   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  0.804877  \n",
      "Sample_13_Ctrl_evening_S13__+\n",
      "There are 48906 unique ids\n",
      "  Adding 103330 missing IDs to sample Sample_13_Ctrl_evening_S13_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      26  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      +      7402              0.003513   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  1.461155  \n",
      "Sample_14_CRS_evening_S14__-\n",
      "There are 49672 unique ids\n",
      "  Adding 102564 missing IDs to sample Sample_14_CRS_evening_S14_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      25  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      -      8786              0.002845   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  1.117473  \n",
      "Sample_14_CRS_evening_S14__+\n",
      "There are 49488 unique ids\n",
      "  Adding 102748 missing IDs to sample Sample_14_CRS_evening_S14_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      18  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      +      7402              0.002432   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  0.955018  \n",
      "Sample_15_Ctrl_evening_S15__-\n",
      "There are 55756 unique ids\n",
      "  Adding 96480 missing IDs to sample Sample_15_Ctrl_evening_S15_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      37  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      -      8786              0.004211   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  1.179539  \n",
      "Sample_15_Ctrl_evening_S15__+\n",
      "There are 55879 unique ids\n",
      "  Adding 96357 missing IDs to sample Sample_15_Ctrl_evening_S15_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      23  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      +      7402              0.003107   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  0.870323  \n",
      "Sample_16_CRS_evening_S16__-\n",
      "There are 52018 unique ids\n",
      "  Adding 100218 missing IDs to sample Sample_16_CRS_evening_S16_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      24  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      -      8786              0.002732   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.948604  \n",
      "Sample_16_CRS_evening_S16__+\n",
      "There are 52247 unique ids\n",
      "  Adding 99989 missing IDs to sample Sample_16_CRS_evening_S16_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      21  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      +      7402              0.002837   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.985225  \n",
      "Sample_17_Ctrl_evening_S17__-\n",
      "There are 51560 unique ids\n",
      "  Adding 100676 missing IDs to sample Sample_17_Ctrl_evening_S17_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      13  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      -      8786               0.00148   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  0.505157  \n",
      "Sample_17_Ctrl_evening_S17__+\n",
      "There are 51670 unique ids\n",
      "  Adding 100566 missing IDs to sample Sample_17_Ctrl_evening_S17_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      22  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      +      7402              0.002972   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  1.014723  \n",
      "Sample_18_CRS_evening_S18__-\n",
      "There are 57265 unique ids\n",
      "  Adding 94971 missing IDs to sample Sample_18_CRS_evening_S18_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      49  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      -      8786              0.005577   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  1.211865  \n",
      "Sample_18_CRS_evening_S18__+\n",
      "There are 57365 unique ids\n",
      "  Adding 94871 missing IDs to sample Sample_18_CRS_evening_S18_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      27  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      +      7402              0.003648   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  0.792618  \n",
      "Sample_19_Ctrl_evening_S19__-\n",
      "There are 50389 unique ids\n",
      "  Adding 101847 missing IDs to sample Sample_19_Ctrl_evening_S19_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-      29  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      -      8786              0.003301  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.296479  \n",
      "Sample_19_Ctrl_evening_S19__+\n",
      "There are 50230 unique ids\n",
      "  Adding 102006 missing IDs to sample Sample_19_Ctrl_evening_S19_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      19  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      +      7402              0.002567  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.008238  \n",
      "Sample_20_CRS_evening_S20__-\n",
      "There are 67231 unique ids\n",
      "  Adding 85005 missing IDs to sample Sample_20_CRS_evening_S20_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001-     100  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      -      8786              0.011382   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                   1.07035  \n",
      "Sample_20_CRS_evening_S20__+\n",
      "There are 67351 unique ids\n",
      "  Adding 84885 missing IDs to sample Sample_20_CRS_evening_S20_ (strand +)\n",
      "                    id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001+      92  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      +      7402              0.012429   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.168842  \n"
     ]
    }
   ],
   "source": [
    "# Fill in missing IDs for each sample\n",
    "print(\"Adding missing IDs to each sample...\")\n",
    "for sample_key, file_data in sample_data_dict.items():\n",
    "    sample_name = sample_key.rsplit('_', 1)[0]\n",
    "\n",
    "    print (sample_key)\n",
    "    \n",
    "    # Identify missing IDs for this sample\n",
    "    existing_ids = set(file_data['id'])\n",
    "    print (f\"There are {len(existing_ids)} unique ids\")\n",
    "    missing_ids = all_unique_ids - existing_ids\n",
    "    \n",
    "    if missing_ids:\n",
    "        print(f\"  Adding {len(missing_ids)} missing IDs to sample {sample_name} (strand {strand})\")\n",
    "        \n",
    "        # Create rows for missing IDs\n",
    "        missing_rows = []\n",
    "        for missing_id in missing_ids:\n",
    "            # Get the chromosome for this ID\n",
    "            \n",
    "            # Create a row with zeros for all numerical values\n",
    "            missing_row = {\n",
    "                'id': missing_id,\n",
    "                'sample': sample_name,\n",
    "                'gene': missing_id[:-1],\n",
    "                'strand': missing_id[-1],\n",
    "                'GC_count': 0,\n",
    "                'damage': 0,\n",
    "                'GC_normalized_damage': 0,\n",
    "                'chromosome': id_to_chrom[missing_id]\n",
    "            }\n",
    "\n",
    "            \n",
    "            # Add median value if available\n",
    "            if sample_name in median_values:\n",
    "                missing_row['median'] = median_values[sample_name]\n",
    "                missing_row['median_normalized_damage'] = 0\n",
    "            else:\n",
    "                missing_row['median'] = np.nan\n",
    "                missing_row['median_normalized_damage'] = np.nan\n",
    "                \n",
    "            missing_rows.append(missing_row)\n",
    "        \n",
    "        # Add the missing rows to the sample's data\n",
    "        if missing_rows:\n",
    "            missing_df = pd.DataFrame(missing_rows)\n",
    "            file_data = pd.concat([file_data, missing_df], ignore_index=True)\n",
    "            sample_data_dict[sample_key] = file_data\n",
    "       \n",
    "        print (file_data.head(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bed0c7e3-318e-411e-a922-5d34380e88af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.32068806.taekim/ipykernel_2431266/224940275.py:3: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  all_data = pd.concat([all_data, file_data], ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing complete! Combined data saved to ../data_normalized/gene_bodies_Normalized.csv\n",
      "Total records: 6089440\n"
     ]
    }
   ],
   "source": [
    "# Combine all samples into the final DataFrame\n",
    "for sample_key, file_data in sample_data_dict.items():\n",
    "    all_data = pd.concat([all_data, file_data], ignore_index=True)\n",
    "\n",
    "# Save the combined data to a CSV file\n",
    "all_data.to_csv(OUTPUT_FILE, index=False)\n",
    "print(f\"Processing complete! Combined data saved to {OUTPUT_FILE}\")\n",
    "print(f\"Total records: {len(all_data)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200caeaf-d53e-4357-8e5f-82958991df49",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
