{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "82445ade-bd78-434f-a921-91a3e6dd51fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import subprocess\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "26097f1d-4ba4-45f0-83eb-562685966bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path with files outputted from bedtools intersect\n",
    "PATH = \"/cluster/scratch/taekim/data_oxidation/promoters_intersect\" \n",
    "OUTPUT_FILE = \"../data_normalized/promoters_Normalized.csv\"\n",
    "GENOME_PATH = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Navnit/genomes/mouse/GRCm39_NCBI_Bowtie2.fasta\"\n",
    "# Original path with all CpG coordinates\n",
    "ORG_PATH = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/knownGenes_canonTr_2000bp_upstream_TSS_feature_GRCm39_GENCODE.VM36.bed\" \n",
    "\n",
    "# Define column names for the input files\n",
    "INPUT_COLUMNS = [\"Chr1\", \"Start1\", \"End1\", \"Value\", \"MAPQ\", \"Chr2\", \"Start2\", \"End2\", \"Gene\", \"Transcript\", \"ignore\"]\n",
    "\n",
    "# Define columns for the output file\n",
    "OUTPUT_COLUMNS = [\"id\", \"sample\", \"gene\", \"strand\", \"GC_count\", \"damage\", \n",
    "                  \"GC_normalized_damage\", \"median\", \"median_normalized_damage\"]\n",
    "\n",
    "# Define the median values for each sample\n",
    "# Median value for each sample from 100kb bins\n",
    "median_values = {\n",
    "    \"Sample_14_CRS_evening_S14_\": 2.546314,\n",
    "    \"Sample_15_Ctrl_evening_S15_\": 3.570246,\n",
    "    \"Sample_05_Ctrl_morning_S5_\": 6.184096,\n",
    "    \"Sample_01_Ctrl_morning_S1_\": 6.921409,\n",
    "    \"Sample_16_CRS_evening_S16_\": 2.879618,\n",
    "    \"Sample_11_Ctrl_evening_S11_\": 2.485161,\n",
    "    \"Sample_13_Ctrl_evening_S13_\": 2.403964,\n",
    "    \"Sample_08_CRS_morning_S8_\": 5.246539,\n",
    "    \"Sample_20_CRS_evening_S20_\": 10.633666,\n",
    "    \"Sample_18_CRS_evening_S18_\": 4.602043,\n",
    "    \"Sample_17_Ctrl_evening_S17_\": 2.929046,\n",
    "    \"Sample_04_CRS_morning_S4_\": 6.227455,\n",
    "    \"Sample_19_Ctrl_evening_S19_\": 2.545900,\n",
    "    \"Sample_03_Ctrl_morning_S3_\": 9.529751,\n",
    "    \"Sample_02_CRS_morning_S2_\": 7.151634,\n",
    "    \"Sample_09_Ctrl_morning_S9_\": 7.923174,\n",
    "    \"Sample_10_CRS_morning_S10_\": 8.815860,\n",
    "    \"Sample_06_CRS_morning_S6_\": 7.930738,\n",
    "    \"Sample_12_CRS_evening_S12_\": 2.849694,\n",
    "    \"Sample_07_Ctrl_morning_S7_\": 4.924262\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4217a79c-cc57-4af1-ba70-9c55bf81cadc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: /cluster/software/stacks/2024-06/spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-12.2.0/bedtools2-2.31.0-a4obbslkxntgdx2criopqpwx662gcftq/bin/bedtools getfasta -fi /nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Navnit/genomes/mouse/GRCm39_NCBI_Bowtie2.fasta -bed /nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/knownGenes_canonTr_2000bp_upstream_TSS_feature_GRCm39_GENCODE.VM36.bed -bedOut > /cluster/scratch/taekim/data_oxidation/promoters_intersect/promoters.SEQ.bed\n",
      "Command completed successfully\n"
     ]
    }
   ],
   "source": [
    "def run_bedtools_getfasta(bed_file, output_file, genome_path):\n",
    "    \"\"\"Run bedtools getfasta command to get sequence data for a BED file\"\"\"\n",
    "    command = f\"/cluster/software/stacks/2024-06/spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-12.2.0/bedtools2-2.31.0-a4obbslkxntgdx2criopqpwx662gcftq/bin/bedtools getfasta -fi {genome_path} -bed {bed_file} -bedOut > {output_file}\"\n",
    "    print(f\"Running command: {command}\")\n",
    "    \n",
    "    try:\n",
    "        result = subprocess.run(command, shell=True, check=True, \n",
    "                               stdout=subprocess.PIPE, stderr=subprocess.PIPE,\n",
    "                               text=True)\n",
    "        print(\"Command completed successfully\")\n",
    "        return True\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running bedtools: {e}\")\n",
    "        print(f\"STDERR: {e.stderr}\")\n",
    "        return False\n",
    "\n",
    "SEQ_COLUMNS = [\"Chr\", \"Start\", \"End\", \"Gene\", \"idk\", \"ignore\", \"Seq\"]\n",
    "seq_output_file = os.path.join(PATH, \"promoters.SEQ.bed\")\n",
    "# Run getfasta with original CpG coordinate file\n",
    "run_bedtools_getfasta(ORG_PATH, seq_output_file, GENOME_PATH)\n",
    "seq_df = pd.read_csv(seq_output_file, sep=\"\\t\", header=None, names=SEQ_COLUMNS)\n",
    "\n",
    "seq_counts_plus = seq_df.copy()\n",
    "seq_counts_plus['strand_id'] = seq_counts_plus['Gene'] + '+'  # Use Gene for consistency with previous code\n",
    "seq_counts_plus['GC_count'] = seq_counts_plus['Seq'].str.upper().str.count('G')\n",
    "seq_counts_plus = seq_counts_plus.drop_duplicates('strand_id')\n",
    "\n",
    "seq_counts_minus = seq_df.copy()\n",
    "seq_counts_minus['strand_id'] = seq_counts_minus['Gene'] + '-'  # Use Gene for consistency with previous code\n",
    "seq_counts_minus['GC_count'] = seq_counts_minus['Seq'].str.upper().str.count('C')\n",
    "seq_counts_minus = seq_counts_minus.drop_duplicates('strand_id')\n",
    "\n",
    "# Combine both strand data\n",
    "seq_counts = pd.concat([seq_counts_plus, seq_counts_minus])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c5c6ed40-ed50-40cf-9d7f-435a0eadabc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First pass: Collecting all unique IDs...\n",
      "Collected 78202 unique IDs across all samples\n"
     ]
    }
   ],
   "source": [
    "# Collect all files for each sample\n",
    "bed_files = glob.glob(os.path.join(PATH, \"*strand.bed\"))\n",
    "\n",
    "all_unique_ids = set()  # To store all unique IDs across all samples\n",
    "# First pass: Collect all unique IDs across all samples\n",
    "print(\"First pass: Collecting all unique IDs...\")\n",
    "for file_path in sorted(bed_files):\n",
    "    \n",
    "    try:\n",
    "        # Read the file\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS,  on_bad_lines='skip')\n",
    "\n",
    "        # Create the identifier\n",
    "        ids = df['Gene']\n",
    "        \n",
    "        # Add to the set of all unique IDs\n",
    "        all_unique_ids.update(ids.unique())\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"  Error collecting IDs from {file_name}: {str(e)}\")\n",
    "\n",
    "print(f\"Collected {len(all_unique_ids)} unique IDs across all samples\")\n",
    "\n",
    "\n",
    "# Create a dictionary to store chromosome info for each ID\n",
    "id_to_chrom = {}\n",
    "for file_path in sorted(bed_files):\n",
    "    \n",
    "    try:\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS)\n",
    "        # Map each ID to its chromosome\n",
    "        id_df = df.copy()\n",
    "        \n",
    "        id_df['id'] = id_df['Gene'] \n",
    "        \n",
    "        for _, row in id_df.drop_duplicates('id').iterrows():\n",
    "            id_to_chrom[row['id']] = row['Chr1']\n",
    "    except Exception as e:\n",
    "        print(f\"  Error mapping chromosomes: {str(e)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3862cee5-bcc8-4547-a834-be4e82e0f1df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting processing of samples\n",
      "starting\n",
      "Processing file: Sample_01_Ctrl_morning_S1_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1777689\n",
      "File data: 65298\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  0.298511  \n",
      "Processing file: Sample_01_Ctrl_morning_S1_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1048587\n",
      "File data: 65396\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      +       415              0.009639   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  1.392571  \n",
      "Processing file: Sample_02_CRS_morning_S2_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1380610\n",
      "File data: 65422\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      -       484              0.006198  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.866704  \n",
      "Processing file: Sample_02_CRS_morning_S2_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1143146\n",
      "File data: 65683\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      +       415              0.004819  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.673871  \n",
      "Processing file: Sample_03_Ctrl_morning_S3_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1922259\n",
      "File data: 71333\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  0.216807  \n",
      "Processing file: Sample_03_Ctrl_morning_S3_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1598487\n",
      "File data: 71281\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       7  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      +       515              0.013592   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  1.426295  \n",
      "Processing file: Sample_04_CRS_morning_S4_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1525004\n",
      "File data: 64065\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       7  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      -       484              0.014463  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  2.322427  \n",
      "Processing file: Sample_04_CRS_morning_S4_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1113101\n",
      "File data: 64014\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       5  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      +       515              0.009709  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.559022  \n",
      "Processing file: Sample_05_Ctrl_morning_S5_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1964552\n",
      "File data: 62428\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       5  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      -       484              0.010331   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  1.670507  \n",
      "Processing file: Sample_05_Ctrl_morning_S5_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1440118\n",
      "File data: 62290\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      +       415              0.007229   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  1.168953  \n",
      "Processing file: Sample_06_CRS_morning_S6_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1765437\n",
      "File data: 69328\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       8  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      -       484              0.016529  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                   2.08416  \n",
      "Processing file: Sample_06_CRS_morning_S6_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1435575\n",
      "File data: 69358\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      +       415              0.009639  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.215341  \n",
      "Processing file: Sample_07_Ctrl_morning_S7_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1646331\n",
      "File data: 57481\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  0.839158  \n",
      "Processing file: Sample_07_Ctrl_morning_S7_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 1165225\n",
      "File data: 57604\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                   0.48934  \n",
      "Processing file: Sample_08_CRS_morning_S8_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1269721\n",
      "File data: 60480\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      -       484              0.008264  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.575222  \n",
      "Processing file: Sample_08_CRS_morning_S8_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 917085\n",
      "File data: 60266\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      +       415              0.004819  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.918563  \n",
      "Processing file: Sample_09_Ctrl_morning_S9_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 828385\n",
      "File data: 67025\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      -       484              0.008264   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.043075  \n",
      "Processing file: Sample_09_Ctrl_morning_S9_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 597635\n",
      "File data: 66883\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       5  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      +       415              0.012048   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.520627  \n",
      "Processing file: Sample_10_CRS_morning_S10_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 694202\n",
      "File data: 71010\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      -       484              0.004132  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.468727  \n",
      "Processing file: Sample_10_CRS_morning_S10_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 566822\n",
      "File data: 71043\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      +       415              0.007229  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                   0.81999  \n",
      "Processing file: Sample_11_Ctrl_evening_S11_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 649161\n",
      "File data: 41094\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.662762  \n",
      "Processing file: Sample_11_Ctrl_evening_S11_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 378809\n",
      "File data: 41324\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       2  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      +       515              0.003883   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.562673  \n",
      "Processing file: Sample_12_CRS_evening_S12_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 728762\n",
      "File data: 46576\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  0.725031  \n",
      "Processing file: Sample_12_CRS_evening_S12_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 426062\n",
      "File data: 46649\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      +       415              0.009639   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  3.382312  \n",
      "Processing file: Sample_13_Ctrl_evening_S13_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1058242\n",
      "File data: 40138\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       1  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      -       421              0.002375   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  0.988075  \n",
      "Processing file: Sample_13_Ctrl_evening_S13_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 536124\n",
      "File data: 40308\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                   1.00236  \n",
      "Processing file: Sample_14_CRS_evening_S14_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 354115\n",
      "File data: 40943\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       2  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      -       421              0.004751   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  1.865675  \n",
      "Processing file: Sample_14_CRS_evening_S14_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 255799\n",
      "File data: 41173\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  0.946324  \n",
      "Processing file: Sample_15_Ctrl_evening_S15_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1126739\n",
      "File data: 54521\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  1.157408  \n",
      "Processing file: Sample_15_Ctrl_evening_S15_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 650990\n",
      "File data: 54687\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  0.674922  \n",
      "Processing file: Sample_16_CRS_evening_S16_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 551259\n",
      "File data: 45808\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000031       1  ENSMUSG00000000031  NC_000073.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      -       514              0.001946   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.675619  \n",
      "Processing file: Sample_16_CRS_evening_S16_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 368315\n",
      "File data: 45824\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       6  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      +       515               0.01165   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  4.045844  \n",
      "Processing file: Sample_17_Ctrl_evening_S17_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 761503\n",
      "File data: 45585\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  0.705389  \n",
      "Processing file: Sample_17_Ctrl_evening_S17_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 442352\n",
      "File data: 45507\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      +       415              0.007229   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                   2.46801  \n",
      "Processing file: Sample_18_CRS_evening_S18_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 548003\n",
      "File data: 58688\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  0.448956  \n",
      "Processing file: Sample_18_CRS_evening_S18_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 449192\n",
      "File data: 58579\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       5  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      +       515              0.009709   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  2.109658  \n",
      "Processing file: Sample_19_Ctrl_evening_S19_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 628834\n",
      "File data: 42125\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000031       4  ENSMUSG00000000031  NC_000073.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      -       514              0.007782  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  3.056719  \n",
      "Processing file: Sample_19_Ctrl_evening_S19_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 348386\n",
      "File data: 41962\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      +       415               0.00241  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.946478  \n",
      "Processing file: Sample_20_CRS_evening_S20_.GRCh38.p13_G_minus_strand.bed\n",
      "  Records: 1562618\n",
      "File data: 75446\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       7  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      -       484              0.014463   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.360096  \n",
      "Processing file: Sample_20_CRS_evening_S20_.GRCh38.p13_G_plus_strand.bed\n",
      "  Records: 987077\n",
      "File data: 75436\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       6  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      +       415              0.014458   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.359628  \n"
     ]
    }
   ],
   "source": [
    "# Create an empty DataFrame for the final output\n",
    "all_data = pd.DataFrame(columns=OUTPUT_COLUMNS)\n",
    "\n",
    "print(\"Starting processing of samples\")\n",
    "\n",
    "# Process each file\n",
    "sample_data_dict = {}  # Store data for each sample\n",
    "\n",
    "print(\"starting\")\n",
    "for file_path in sorted(bed_files):\n",
    "    file_name = os.path.basename(file_path)\n",
    "    print(f\"Processing file: {file_name}\")\n",
    "    \n",
    "    # Extract sample name and determine strand\n",
    "    sample_name = file_name.split('.')[0]\n",
    "    \n",
    "    if \"plus_strand\" in file_name:\n",
    "        strand = \"+\"\n",
    "    elif \"minus_strand\" in file_name:\n",
    "        strand = \"-\"\n",
    "    else:\n",
    "        print(f\"  Warning: Could not determine strand for {file_name}\")\n",
    "        continue\n",
    "    \n",
    "    try:\n",
    "        # Read and process the file\n",
    "        # dataframe of intersection from oxidation sites per sample and cpg coordinates\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=INPUT_COLUMNS)\n",
    "        print(f\"  Records: {len(df)}\")\n",
    "        \n",
    "        # Create IDs and calculate damage per genomic position\n",
    "        df['id'] = df['Gene'] \n",
    "        df['strand_id'] = df['Gene'] + strand\n",
    "        \n",
    "        # Create damage_df with both summed damage and first gene in one operation\n",
    "        damage_df = df.groupby('id').agg({\n",
    "            'Value': 'sum',\n",
    "            'Gene': 'first',\n",
    "            'Chr1': 'first',\n",
    "            'strand_id': 'first'\n",
    "        }).reset_index()\n",
    "        \n",
    "        # Rename Value column to damage\n",
    "        damage_df = damage_df.rename(columns={'Value': 'damage'})\n",
    "        damage_df = damage_df.rename(columns={'Gene': 'gene'})\n",
    "        damage_df = damage_df.rename(columns={'Chr1': 'chromosome'})\n",
    "        \n",
    "        # Create base result dataframe\n",
    "        file_data = damage_df.copy()\n",
    "        file_data['sample'] = sample_name\n",
    "        file_data['strand'] = strand\n",
    "        \n",
    "        # 4. Merge the counts with our result dataframe\n",
    "        file_data = file_data.merge(\n",
    "            seq_counts[['strand_id', 'GC_count']], \n",
    "            on='strand_id', \n",
    "            how='left'\n",
    "        )\n",
    "        file_data.drop('strand_id', axis=1, inplace=True)\n",
    "        \n",
    "        # Calculate normalized damage\n",
    "        file_data['GC_normalized_damage'] = file_data['damage'] / file_data['GC_count'].replace(0, np.nan)\n",
    "        \n",
    "        # Add median normalization if available\n",
    "        if sample_name in median_values:\n",
    "            file_data['median'] = median_values[sample_name]\n",
    "            file_data['median_normalized_damage'] = 1000 * file_data['GC_normalized_damage'] / file_data['median']\n",
    "        else:\n",
    "            print(f\"  Warning: No median value found for sample {sample_name}\")\n",
    "            file_data['median'] = np.nan\n",
    "            file_data['median_normalized_damage'] = np.nan\n",
    "        \n",
    "        # Store processed data\n",
    "        # file_data saved per sample\n",
    "        sample_data_dict[f\"{sample_name}_{strand}\"] = file_data\n",
    "        print(f\"File data: {len(file_data)}\")\n",
    "        print (file_data.head(1))\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"  Error processing {file_name}: {str(e)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6b34f245-69af-449e-be0c-b9d8f698cfd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding missing IDs to each sample...\n",
      "There are 65298 unique ids\n",
      "  Adding 12904 missing IDs to sample Sample_01_Ctrl_morning_S1_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  0.298511  \n",
      "There are 65396 unique ids\n",
      "  Adding 12806 missing IDs to sample Sample_01_Ctrl_morning_S1_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_01_Ctrl_morning_S1_      +       415              0.009639   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.921409                  1.392571  \n",
      "There are 65422 unique ids\n",
      "  Adding 12780 missing IDs to sample Sample_02_CRS_morning_S2_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      -       484              0.006198  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.866704  \n",
      "There are 65683 unique ids\n",
      "  Adding 12519 missing IDs to sample Sample_02_CRS_morning_S2_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_02_CRS_morning_S2_      +       415              0.004819  7.151634   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.673871  \n",
      "There are 71333 unique ids\n",
      "  Adding 6869 missing IDs to sample Sample_03_Ctrl_morning_S3_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  0.216807  \n",
      "There are 71281 unique ids\n",
      "  Adding 6921 missing IDs to sample Sample_03_Ctrl_morning_S3_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       7  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_03_Ctrl_morning_S3_      +       515              0.013592   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  9.529751                  1.426295  \n",
      "There are 64065 unique ids\n",
      "  Adding 14137 missing IDs to sample Sample_04_CRS_morning_S4_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       7  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      -       484              0.014463  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  2.322427  \n",
      "There are 64014 unique ids\n",
      "  Adding 14188 missing IDs to sample Sample_04_CRS_morning_S4_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       5  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_04_CRS_morning_S4_      +       515              0.009709  6.227455   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.559022  \n",
      "There are 62428 unique ids\n",
      "  Adding 15774 missing IDs to sample Sample_05_Ctrl_morning_S5_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       5  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      -       484              0.010331   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  1.670507  \n",
      "There are 62290 unique ids\n",
      "  Adding 15912 missing IDs to sample Sample_05_Ctrl_morning_S5_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_05_Ctrl_morning_S5_      +       415              0.007229   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  6.184096                  1.168953  \n",
      "There are 69328 unique ids\n",
      "  Adding 8874 missing IDs to sample Sample_06_CRS_morning_S6_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       8  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      -       484              0.016529  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                   2.08416  \n",
      "There are 69358 unique ids\n",
      "  Adding 8844 missing IDs to sample Sample_06_CRS_morning_S6_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_06_CRS_morning_S6_      +       415              0.009639  7.930738   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.215341  \n",
      "There are 57481 unique ids\n",
      "  Adding 20721 missing IDs to sample Sample_07_Ctrl_morning_S7_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                  0.839158  \n",
      "There are 57604 unique ids\n",
      "  Adding 20598 missing IDs to sample Sample_07_Ctrl_morning_S7_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_07_Ctrl_morning_S7_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.924262                   0.48934  \n",
      "There are 60480 unique ids\n",
      "  Adding 17722 missing IDs to sample Sample_08_CRS_morning_S8_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      -       484              0.008264  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  1.575222  \n",
      "There are 60266 unique ids\n",
      "  Adding 17936 missing IDs to sample Sample_08_CRS_morning_S8_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                      sample strand  GC_count  GC_normalized_damage    median  \\\n",
      "0  Sample_08_CRS_morning_S8_      +       415              0.004819  5.246539   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.918563  \n",
      "There are 67025 unique ids\n",
      "  Adding 11177 missing IDs to sample Sample_09_Ctrl_morning_S9_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      -       484              0.008264   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.043075  \n",
      "There are 66883 unique ids\n",
      "  Adding 11319 missing IDs to sample Sample_09_Ctrl_morning_S9_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       5  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_09_Ctrl_morning_S9_      +       415              0.012048   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  7.923174                  1.520627  \n",
      "There are 71010 unique ids\n",
      "  Adding 7192 missing IDs to sample Sample_10_CRS_morning_S10_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      -       484              0.004132  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.468727  \n",
      "There are 71043 unique ids\n",
      "  Adding 7159 missing IDs to sample Sample_10_CRS_morning_S10_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage   median  \\\n",
      "0  Sample_10_CRS_morning_S10_      +       415              0.007229  8.81586   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                   0.81999  \n",
      "There are 41094 unique ids\n",
      "  Adding 37108 missing IDs to sample Sample_11_Ctrl_evening_S11_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.662762  \n",
      "There are 41324 unique ids\n",
      "  Adding 36878 missing IDs to sample Sample_11_Ctrl_evening_S11_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       2  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_11_Ctrl_evening_S11_      +       515              0.003883   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.485161                  1.562673  \n",
      "There are 46576 unique ids\n",
      "  Adding 31626 missing IDs to sample Sample_12_CRS_evening_S12_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  0.725031  \n",
      "There are 46649 unique ids\n",
      "  Adding 31553 missing IDs to sample Sample_12_CRS_evening_S12_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       4  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_12_CRS_evening_S12_      +       415              0.009639   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.849694                  3.382312  \n",
      "There are 40138 unique ids\n",
      "  Adding 38064 missing IDs to sample Sample_13_Ctrl_evening_S13_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       1  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      -       421              0.002375   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                  0.988075  \n",
      "There are 40308 unique ids\n",
      "  Adding 37894 missing IDs to sample Sample_13_Ctrl_evening_S13_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_13_Ctrl_evening_S13_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.403964                   1.00236  \n",
      "There are 40943 unique ids\n",
      "  Adding 37259 missing IDs to sample Sample_14_CRS_evening_S14_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       2  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      -       421              0.004751   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  1.865675  \n",
      "There are 41173 unique ids\n",
      "  Adding 37029 missing IDs to sample Sample_14_CRS_evening_S14_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_14_CRS_evening_S14_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.546314                  0.946324  \n",
      "There are 54521 unique ids\n",
      "  Adding 23681 missing IDs to sample Sample_15_Ctrl_evening_S15_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       2  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      -       484              0.004132   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  1.157408  \n",
      "There are 54687 unique ids\n",
      "  Adding 23515 missing IDs to sample Sample_15_Ctrl_evening_S15_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_15_Ctrl_evening_S15_      +       415               0.00241   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  3.570246                  0.674922  \n",
      "There are 45808 unique ids\n",
      "  Adding 32394 missing IDs to sample Sample_16_CRS_evening_S16_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000031       1  ENSMUSG00000000031  NC_000073.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      -       514              0.001946   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  0.675619  \n",
      "There are 45824 unique ids\n",
      "  Adding 32378 missing IDs to sample Sample_16_CRS_evening_S16_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       6  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_16_CRS_evening_S16_      +       515               0.01165   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.879618                  4.045844  \n",
      "There are 45585 unique ids\n",
      "  Adding 32617 missing IDs to sample Sample_17_Ctrl_evening_S17_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                  0.705389  \n",
      "There are 45507 unique ids\n",
      "  Adding 32695 missing IDs to sample Sample_17_Ctrl_evening_S17_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       3  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_17_Ctrl_evening_S17_      +       415              0.007229   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  2.929046                   2.46801  \n",
      "There are 58688 unique ids\n",
      "  Adding 19514 missing IDs to sample Sample_18_CRS_evening_S18_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      -       484              0.002066   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  0.448956  \n",
      "There are 58579 unique ids\n",
      "  Adding 19623 missing IDs to sample Sample_18_CRS_evening_S18_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000028       5  ENSMUSG00000000028  NC_000082.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_18_CRS_evening_S18_      +       515              0.009709   \n",
      "\n",
      "     median  median_normalized_damage  \n",
      "0  4.602043                  2.109658  \n",
      "There are 42125 unique ids\n",
      "  Adding 36077 missing IDs to sample Sample_19_Ctrl_evening_S19_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000031       4  ENSMUSG00000000031  NC_000073.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      -       514              0.007782  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  3.056719  \n",
      "There are 41962 unique ids\n",
      "  Adding 36240 missing IDs to sample Sample_19_Ctrl_evening_S19_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       1  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                        sample strand  GC_count  GC_normalized_damage  median  \\\n",
      "0  Sample_19_Ctrl_evening_S19_      +       415               0.00241  2.5459   \n",
      "\n",
      "   median_normalized_damage  \n",
      "0                  0.946478  \n",
      "There are 75446 unique ids\n",
      "  Adding 2756 missing IDs to sample Sample_20_CRS_evening_S20_ (strand -)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       7  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      -       484              0.014463   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.360096  \n",
      "There are 75436 unique ids\n",
      "  Adding 2766 missing IDs to sample Sample_20_CRS_evening_S20_ (strand +)\n",
      "                   id  damage                gene   chromosome  \\\n",
      "0  ENSMUSG00000000001       6  ENSMUSG00000000001  NC_000069.7   \n",
      "\n",
      "                       sample strand  GC_count  GC_normalized_damage  \\\n",
      "0  Sample_20_CRS_evening_S20_      +       415              0.014458   \n",
      "\n",
      "      median  median_normalized_damage  \n",
      "0  10.633666                  1.359628  \n"
     ]
    }
   ],
   "source": [
    "# Fill in missing IDs for each sample\n",
    "print(\"Adding missing IDs to each sample...\")\n",
    "for sample_key, file_data in sample_data_dict.items():\n",
    "    sample_name = sample_key.rsplit('_', 1)[0]\n",
    "    strand = sample_key.split('_')[-1]\n",
    "    \n",
    "    # Identify missing IDs for this sample\n",
    "    existing_ids = set(file_data['id'])\n",
    "    print (f\"There are {len(existing_ids)} unique ids\")\n",
    "    missing_ids = all_unique_ids - existing_ids\n",
    "    \n",
    "    if missing_ids:\n",
    "        print(f\"  Adding {len(missing_ids)} missing IDs to sample {sample_name} (strand {strand})\")\n",
    "        \n",
    "        # Create rows for missing IDs\n",
    "        missing_rows = []\n",
    "        for missing_id in missing_ids:\n",
    "            # Get the chromosome for this ID\n",
    "            \n",
    "            # Create a row with zeros for all numerical values\n",
    "            missing_row = {\n",
    "                'id': missing_id,\n",
    "                'sample': sample_name,\n",
    "                'gene': missing_id,\n",
    "                'strand': strand,\n",
    "                'GC_count': 0,\n",
    "                'damage': 0,\n",
    "                'GC_normalized_damage': 0,\n",
    "                'chromosome': id_to_chrom[missing_id]\n",
    "            }\n",
    "\n",
    "            \n",
    "            # Add median value if available\n",
    "            if sample_name in median_values:\n",
    "                missing_row['median'] = median_values[sample_name]\n",
    "                missing_row['median_normalized_damage'] = 0\n",
    "            else:\n",
    "                missing_row['median'] = np.nan\n",
    "                missing_row['median_normalized_damage'] = np.nan\n",
    "                \n",
    "            missing_rows.append(missing_row)\n",
    "        \n",
    "        # Add the missing rows to the sample's data\n",
    "        if missing_rows:\n",
    "            missing_df = pd.DataFrame(missing_rows)\n",
    "            file_data = pd.concat([file_data, missing_df], ignore_index=True)\n",
    "            sample_data_dict[sample_key] = file_data\n",
    "       \n",
    "        print (file_data.head(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bed0c7e3-318e-411e-a922-5d34380e88af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.32115401.taekim/ipykernel_3032713/224940275.py:3: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  all_data = pd.concat([all_data, file_data], ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing complete! Combined data saved to ../data_normalized/promoters_Normalized.csv\n",
      "Total records: 3128080\n"
     ]
    }
   ],
   "source": [
    "# Combine all samples into the final DataFrame\n",
    "for sample_key, file_data in sample_data_dict.items():\n",
    "    all_data = pd.concat([all_data, file_data], ignore_index=True)\n",
    "\n",
    "# Save the combined data to a CSV file\n",
    "all_data.to_csv(OUTPUT_FILE, index=False)\n",
    "print(f\"Processing complete! Combined data saved to {OUTPUT_FILE}\")\n",
    "print(f\"Total records: {len(all_data)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200caeaf-d53e-4357-8e5f-82958991df49",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
