{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.11.6 (main, Jun  7 2024, 07:09:59) [GCC 13.2.0]\n",
      "numpy 1.26.4\n",
      "pandas 2.2.2\n"
     ]
    }
   ],
   "source": [
    "#importing necessary modules\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "print(sys.version)\n",
    "print(\"numpy\", np.__version__)\n",
    "print(\"pandas\", pd.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1',\n",
       " 'chr2',\n",
       " 'chr3',\n",
       " 'chr4',\n",
       " 'chr5',\n",
       " 'chr6',\n",
       " 'chr7',\n",
       " 'chr8',\n",
       " 'chr9',\n",
       " 'chr10',\n",
       " 'chr11',\n",
       " 'chr12',\n",
       " 'chr13',\n",
       " 'chr14',\n",
       " 'chr15',\n",
       " 'chr16',\n",
       " 'chr17',\n",
       " 'chr18',\n",
       " 'chr19',\n",
       " 'chrX',\n",
       " 'chrY']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosomes = ['chr' + str(i) for i in np.arange(1, 20, 1)] + [\"chrX\", \"chrY\"]\n",
    "chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RefSeq seq accession</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000068.8</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000069.7</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000070.7</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000071.7</td>\n",
       "      <td>chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000072.7</td>\n",
       "      <td>chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000073.7</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_000074.7</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_000075.7</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_000076.7</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_000077.7</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_000078.7</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_000079.7</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_000080.7</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_000081.7</td>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_000082.7</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_000083.7</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_000084.7</td>\n",
       "      <td>chr18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_000085.7</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   RefSeq seq accession  chrom\n",
       "0           NC_000067.7   chr1\n",
       "1           NC_000068.8   chr2\n",
       "2           NC_000069.7   chr3\n",
       "3           NC_000070.7   chr4\n",
       "4           NC_000071.7   chr5\n",
       "5           NC_000072.7   chr6\n",
       "6           NC_000073.7   chr7\n",
       "7           NC_000074.7   chr8\n",
       "8           NC_000075.7   chr9\n",
       "9           NC_000076.7  chr10\n",
       "10          NC_000077.7  chr11\n",
       "11          NC_000078.7  chr12\n",
       "12          NC_000079.7  chr13\n",
       "13          NC_000080.7  chr14\n",
       "14          NC_000081.7  chr15\n",
       "15          NC_000082.7  chr16\n",
       "16          NC_000083.7  chr17\n",
       "17          NC_000084.7  chr18\n",
       "18          NC_000085.7  chr19\n",
       "19          NC_000086.8   chrX\n",
       "20          NC_000087.8   chrY"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosomes_name_table = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Navnit/genomes/mouse/Chromosome_naming_table_17.09.2024_Vakil.tsv\"\n",
    "chromosomes_name_table = pd.read_csv(chromosomes_name_table, sep = \"\\t\", header = 0)\n",
    "chromosomes_name_table = chromosomes_name_table[chromosomes_name_table[\"UCSC style name\"].isin(chromosomes)]\n",
    "chromosomes_name_table = chromosomes_name_table.loc[:, [\"RefSeq seq accession\", \"UCSC style name\"]]\n",
    "chromosomes_name_table = chromosomes_name_table.rename(columns = {\"UCSC style name\" : \"chrom\"})\n",
    "chromosomes_name_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>3143475</td>\n",
       "      <td>3144545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000082908.3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>3172238</td>\n",
       "      <td>3172348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3276123</td>\n",
       "      <td>3286567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000159265.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3276745</td>\n",
       "      <td>3285855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000070533.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3284704</td>\n",
       "      <td>3741721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278370</th>\n",
       "      <td>ENSMUST00000180208.2</td>\n",
       "      <td>chr5_JH584299v1_random</td>\n",
       "      <td>-</td>\n",
       "      <td>910288</td>\n",
       "      <td>913083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278371</th>\n",
       "      <td>ENSMUST00000178327.2</td>\n",
       "      <td>chr5_JH584299v1_random</td>\n",
       "      <td>+</td>\n",
       "      <td>921941</td>\n",
       "      <td>924675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278372</th>\n",
       "      <td>ENSMUST00000177783.2</td>\n",
       "      <td>chr7_GL456219v1_random</td>\n",
       "      <td>-</td>\n",
       "      <td>94185</td>\n",
       "      <td>95088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278373</th>\n",
       "      <td>ENSMUST00000177979.2</td>\n",
       "      <td>chr7_GL456219v1_random</td>\n",
       "      <td>-</td>\n",
       "      <td>149260</td>\n",
       "      <td>150163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278374</th>\n",
       "      <td>ENSMUST00000178569.2</td>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>+</td>\n",
       "      <td>81606</td>\n",
       "      <td>82689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>278375 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name                   chrom strand  chromStart  \\\n",
       "0       ENSMUST00000193812.2                    chr1      +     3143475   \n",
       "1       ENSMUST00000082908.3                    chr1      +     3172238   \n",
       "2       ENSMUST00000162897.2                    chr1      -     3276123   \n",
       "3       ENSMUST00000159265.2                    chr1      -     3276745   \n",
       "4       ENSMUST00000070533.5                    chr1      -     3284704   \n",
       "...                      ...                     ...    ...         ...   \n",
       "278370  ENSMUST00000180208.2  chr5_JH584299v1_random      -      910288   \n",
       "278371  ENSMUST00000178327.2  chr5_JH584299v1_random      +      921941   \n",
       "278372  ENSMUST00000177783.2  chr7_GL456219v1_random      -       94185   \n",
       "278373  ENSMUST00000177979.2  chr7_GL456219v1_random      -      149260   \n",
       "278374  ENSMUST00000178569.2  chrY_JH584303v1_random      +       81606   \n",
       "\n",
       "        chromEnd  \n",
       "0        3144545  \n",
       "1        3172348  \n",
       "2        3286567  \n",
       "3        3285855  \n",
       "4        3741721  \n",
       "...          ...  \n",
       "278370    913083  \n",
       "278371    924675  \n",
       "278372     95088  \n",
       "278373    150163  \n",
       "278374     82689  \n",
       "\n",
       "[278375 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/UCSC_Table_Browser/knownGene_GRCm39_GENCODE.VM36.tsv\"\n",
    "allGenecode = pd.read_csv(path, sep = \"\\t\")\n",
    "allGenecode = allGenecode.loc[:, [\"name\", \"#chrom\", \"strand\", \"chromStart\", \"chromEnd\"]]\n",
    "allGenecode = allGenecode.rename(columns = {\"#chrom\" : \"chrom\"})\n",
    "allGenecode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78202, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>#name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>165957806</td>\n",
       "      <td>165994079</td>\n",
       "      <td>ENSMUSG00000000544</td>\n",
       "      <td>ENSMUST00000060833.14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>161608257</td>\n",
       "      <td>161616064</td>\n",
       "      <td>ENSMUSG00000000817</td>\n",
       "      <td>ENSMUST00000000834.4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>36550947</td>\n",
       "      <td>36567318</td>\n",
       "      <td>ENSMUSG00000001138</td>\n",
       "      <td>ENSMUST00000001166.14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>36461145</td>\n",
       "      <td>36484352</td>\n",
       "      <td>ENSMUSG00000001143</td>\n",
       "      <td>ENSMUST00000125304.8</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>186453174</td>\n",
       "      <td>186481555</td>\n",
       "      <td>ENSMUSG00000001305</td>\n",
       "      <td>ENSMUST00000001339.6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78197</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>85425286</td>\n",
       "      <td>85425394</td>\n",
       "      <td>ENSMUSG00002076147</td>\n",
       "      <td>ENSMUST00020182676.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78198</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>83823856</td>\n",
       "      <td>83823964</td>\n",
       "      <td>ENSMUSG00002076259</td>\n",
       "      <td>ENSMUST00020182612.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78199</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>86337285</td>\n",
       "      <td>86337393</td>\n",
       "      <td>ENSMUSG00002076549</td>\n",
       "      <td>ENSMUST00020181692.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78200</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>54925575</td>\n",
       "      <td>54925683</td>\n",
       "      <td>ENSMUSG00002076770</td>\n",
       "      <td>ENSMUST00020183724.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78201</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>991656</td>\n",
       "      <td>991716</td>\n",
       "      <td>ENSMUSG00002076920</td>\n",
       "      <td>ENSMUST00020183589.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78202 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             chrom    txStart      txEnd             Gene ID  \\\n",
       "0      NC_000067.7  165957806  165994079  ENSMUSG00000000544   \n",
       "1      NC_000067.7  161608257  161616064  ENSMUSG00000000817   \n",
       "2      NC_000067.7   36550947   36567318  ENSMUSG00000001138   \n",
       "3      NC_000067.7   36461145   36484352  ENSMUSG00000001143   \n",
       "4      NC_000067.7  186453174  186481555  ENSMUSG00000001305   \n",
       "...            ...        ...        ...                 ...   \n",
       "78197  NC_000087.8   85425286   85425394  ENSMUSG00002076147   \n",
       "78198  NC_000087.8   83823856   83823964  ENSMUSG00002076259   \n",
       "78199  NC_000087.8   86337285   86337393  ENSMUSG00002076549   \n",
       "78200  NC_000087.8   54925575   54925683  ENSMUSG00002076770   \n",
       "78201  NC_000087.8     991656     991716  ENSMUSG00002076920   \n",
       "\n",
       "                       #name strand  \n",
       "0      ENSMUST00000060833.14      +  \n",
       "1       ENSMUST00000000834.4      -  \n",
       "2      ENSMUST00000001166.14      +  \n",
       "3       ENSMUST00000125304.8      -  \n",
       "4       ENSMUST00000001339.6      -  \n",
       "...                      ...    ...  \n",
       "78197   ENSMUST00020182676.1      -  \n",
       "78198   ENSMUST00020182612.1      -  \n",
       "78199   ENSMUST00020181692.1      -  \n",
       "78200   ENSMUST00020183724.1      -  \n",
       "78201   ENSMUST00020183589.1      -  \n",
       "\n",
       "[78202 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/UCSC_Table_Browser/knownCanonical_Genes_GRCm39_GENCODE.VM36.tsv\"\n",
    "canonGenecode = pd.read_csv(path, sep = \"\\t\")\n",
    "canonGenecode[\"Gene ID\"] = canonGenecode[\"protein\"].str.split(\".\").str[0]\n",
    "#\n",
    "canonGenecode = canonGenecode[canonGenecode[\"#chrom\"].isin(chromosomes)]\n",
    "#\n",
    "canonGenecode = canonGenecode.loc[:, [\"transcript\", \"Gene ID\"]]\n",
    "canonGenecode = canonGenecode.rename(columns = {\"transcript\" : \"name\"})\n",
    "print(canonGenecode.shape)\n",
    "\n",
    "canonGenecode = pd.merge(canonGenecode, allGenecode, on = \"name\", how = \"left\")\n",
    "canonGenecode = pd.merge(chromosomes_name_table, canonGenecode, on = \"chrom\")\n",
    "canonGenecode = canonGenecode.loc[:, [\"RefSeq seq accession\", \"chromStart\", \"chromEnd\", \"Gene ID\", \"name\", \"strand\"]]\n",
    "canonGenecode = canonGenecode.rename(columns = {\"RefSeq seq accession\" : \"chrom\",\n",
    "                                                 \"chromStart\" : \"txStart\",\n",
    "                                                 \"chromEnd\" : \"txEnd\",\n",
    "                                                 \"name\" : \"#name\"})\n",
    "\n",
    "canonGenecode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#name\n",
       "ENSMUST00000060833.14    1\n",
       "ENSMUST00000246328.1     1\n",
       "ENSMUST00000362961.1     1\n",
       "ENSMUST00000247074.1     1\n",
       "ENSMUST00000318361.1     1\n",
       "                        ..\n",
       "ENSMUST00000204200.2     1\n",
       "ENSMUST00000164674.3     1\n",
       "ENSMUST00000204643.2     1\n",
       "ENSMUST00000203812.2     1\n",
       "ENSMUST00020183589.1     1\n",
       "Name: count, Length: 78202, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "canonGenecode[\"#name\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID\n",
       "ENSMUSG00000000544    1\n",
       "ENSMUSG00000120893    1\n",
       "ENSMUSG00000121126    1\n",
       "ENSMUSG00000121073    1\n",
       "ENSMUSG00000121052    1\n",
       "                     ..\n",
       "ENSMUSG00000107823    1\n",
       "ENSMUSG00000107821    1\n",
       "ENSMUSG00000107819    1\n",
       "ENSMUSG00000107818    1\n",
       "ENSMUSG00002076920    1\n",
       "Name: count, Length: 78202, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "canonGenecode[\"Gene ID\"].value_counts() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "P_drive_path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/\"\n",
    "\n",
    "canonGenecode.to_csv(P_drive_path + \"knownGenes_canononical_GRCm39_GENCODE.VM36.bed\", sep = \"\\t\", index = False, header = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "P_drive_path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/\"\n",
    "\n",
    "beyond = 2000\n",
    "\n",
    "#upstream TSS region\n",
    "DF_upstream_bounds = pd.DataFrame({})\n",
    "for strand in [\"-\", \"+\"]:\n",
    "    tmp = canonGenecode[canonGenecode[\"strand\"] == strand].copy()\n",
    "    if strand == \"+\":\n",
    "        tmp.loc[:, \"End\"] = tmp[\"txStart\"]\n",
    "        tmp.loc[:, \"Start\"] = tmp[\"txStart\"] - beyond\n",
    "    if strand == \"-\":\n",
    "        tmp.loc[:, \"Start\"] = tmp[\"txEnd\"]\n",
    "        tmp.loc[:, \"End\"] = tmp[\"txEnd\"] + beyond\n",
    "        \n",
    "    tmp = tmp.loc[:, [\"chrom\", \"Start\", \"End\", \"Gene ID\", \"#name\", \"strand\"]]\n",
    "    DF_upstream_bounds = pd.concat([DF_upstream_bounds, tmp])\n",
    "    \n",
    "DF_upstream_bounds.to_csv(P_drive_path + \"knownGenes_canonTr_\"+str(beyond)+\"bp_upstream_TSS_feature_GRCm39_GENCODE.VM36.bed\", \n",
    "                          header = False, index = False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RefSeq seq accession</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>91225617</td>\n",
       "      <td>91226516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>38665874</td>\n",
       "      <td>38667165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>57016263</td>\n",
       "      <td>57016551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>85721048</td>\n",
       "      <td>85722212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>88997698</td>\n",
       "      <td>88998967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15953</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>90795245</td>\n",
       "      <td>90797512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15954</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>90804558</td>\n",
       "      <td>90805088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15955</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>90829598</td>\n",
       "      <td>90829834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15956</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>90835535</td>\n",
       "      <td>90835771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15957</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>90836409</td>\n",
       "      <td>90840591</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15958 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RefSeq seq accession  chromStart  chromEnd\n",
       "0              NC_000067.7    91225617  91226516\n",
       "1              NC_000067.7    38665874  38667165\n",
       "2              NC_000067.7    57016263  57016551\n",
       "3              NC_000067.7    85721048  85722212\n",
       "4              NC_000067.7    88997698  88998967\n",
       "...                    ...         ...       ...\n",
       "15953          NC_000087.8    90795245  90797512\n",
       "15954          NC_000087.8    90804558  90805088\n",
       "15955          NC_000087.8    90829598  90829834\n",
       "15956          NC_000087.8    90835535  90835771\n",
       "15957          NC_000087.8    90836409  90840591\n",
       "\n",
       "[15958 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/UCSC_Table_Browser/CpG_islands_GRCm39.tsv\"\n",
    "CpG_islands = pd.read_csv(path, sep = \"\\t\")\n",
    "CpG_islands = CpG_islands.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]]\n",
    "CpG_islands = pd.merge(chromosomes_name_table, CpG_islands, on = \"chrom\")\n",
    "CpG_islands = CpG_islands.loc[:, [\"RefSeq seq accession\", \"chromStart\", \"chromEnd\"]]\n",
    "\n",
    "P_drive_path = \"/nfs/nas12.ethz.ch/fs1201/green_groups_let_public/Euler/Vakil/mouse_genome_annotation/Genes_Promoters_CpG_islands_for_Tae/\"\n",
    "CpG_islands.to_csv(P_drive_path + \"allCpG_islands_GRCm39.bed\", \n",
    "                          header = False, index = False, sep = '\\t')\n",
    "CpG_islands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
